compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs

```diff
@@ -281,3 +283,39 @@ getPackageMetadata :: Version -> PackageName -> Maybe PackageVersion -> Maybe Pa
 getPackageMetadata lfVer pkgName mbPkgVersion = do
     guard (lfVer `supports` featurePackageMetadata)
     Just (PackageMetadata pkgName (fromMaybe (PackageVersion "0.0.0") mbPkgVersion))
+
+-- | Given the name of a DALF and the decoded package return package metadata.
+--
+-- For newer DAML-LF versions this is taken directly from the
+-- package metadata in DAML-LF. For older versions, we instead infer
+-- metadata from the filename.
+packageMetadataFromFile :: FilePath -> Package -> PackageId -> (PackageName, Maybe PackageVersion)
+packageMetadataFromFile file pkg pkgId
+    | Just (PackageMetadata name version) <- packageMetadata pkg =
+          -- GHC insists on daml-prim not having a version so we filter it out.
+          (name, version <$ guard (name /= PackageName "daml-prim"))
+    | otherwise = splitUnitId (unitIdFromFile file pkgId)
+
+-- Get the name of a file and an expeted package id of the package, get the unit id
+-- by stripping away the package name at the end.
+-- E.g., if 'package-name-123abc' is given and the known package id is
+-- '123abc', then 'package-name' is returned as unit id.
+unitIdFromFile :: FilePath -> PackageId -> UnitId
+unitIdFromFile file (PackageId pkgId) =
+    (stringToUnitId . fromMaybe name . stripPkgId name . T.unpack) pkgId
+    where name = takeBaseName file
+
+-- Strip the package id from the end of a dalf file name
+-- TODO (drsk) This needs to become a hard error
+stripPkgId :: String -> String -> Maybe String
+stripPkgId baseName expectedPkgId = do
+    (unitId, pkgId) <- stripInfixEnd "-" baseName
+    guard $ pkgId == expectedPkgId
+    pure unitId
+
+-- | Take a string of the form daml-stdlib-"0.13.43" and split it into ("daml-stdlib", Just "0.13.43")
+splitUnitId :: UnitId -> (PackageName, Maybe PackageVersion)
```

Comment:
Perhaps we would be better turning `PackageMetadata`'s package version field into a `Maybe`, given this function, the `packageMetadataFromFile` function, and the arbitrary version 0.0.0 being defaulted above in `getPackageMetadata`. Then we can just return a `PackageMetadata` here and in `packageMetadataFromFile`.

Reply:
I’m a bit unsure about this. The only case where we really don’t have a version is `daml-prim`. I’m not sure we want to make the version optional in LF just to be able to handle this case.

Reply:
You know more about it than me, so I'm happy to go with whatever way you think works best :-)

Reply:
Alright, then I suggest to keep it like this for now to not leak GHC “defficiencies” into DAML-LF. As mentioned, I’d like to cleanup the whole code around `daml-prim`/`dam-stdlib` a bit and as part of that we can probably also put in more stricter enforcement that `daml-prim` is the only thing without a version (maybe something like `data DamlPackageMetadata = DamlPrim | DamlPackageMetadata PackageName PackageVersion`).